### ISSUE TITLE
Look into Singularity support

### ISSUE THREAD

**Issue by epicfaace** [codalab/codalab-worksheets#3284]
Make a doc showing what we would need to change in order to support Singularity (in addition to Docker) for worker runs.

Motivation: We want to use Singularity instead of Docker on the NLP cluster (because Docker requires root access).

https://docs.google.com/document/d/1QfH3xk84xByjrK19BwcbrZ87_F2LJe4MAbts5hx4nBE/edit

**Comment by epicfaace** [comment]
Try running codalab default-cpu / default-gpu images with Singularity (they're around 5 GB large) with other images from different bundles. Try a list of ~100 images and see how many of them work in Singularity.

**Comment by epicfaace** [comment]
Test with these images: https://github.com/codalab/codalab-worksheets/blob/a01cf12efa1f36c70577c8c74cb0460ae2fb100b/tests/stress/stress_test.py#L83-L102

Update design doc to describe implementation (which classes would change, in what way, how much work)

**Comment by epicfaace** [comment]
- Talk to users of singularity
- Think about API changes (--request-docker-image -> --request-singularity-image), caching, etc.

**Comment by jzwang43** [comment]
> 
> 
> Test with these images:
> 
> https://github.com/codalab/codalab-worksheets/blob/a01cf12efa1f36c70577c8c74cb0460ae2fb100b/tests/stress/stress_test.py#L83-L102
> 
> Update design doc to describe implementation (which classes would change, in what way, how much work)

I have tested the images listed and looks like `mongo:latest` `large/rdp:dev`  `neo4j`  `jenkins/jenkins` don't seem to work out of the box with singularity. `codalab cpu` and `codalab gpu` worked but were very slow due to large image size.

**Comment by adiprerepa** [comment]
Here are the steps we need to take:
- [ ] Abstract ImageManager into its base class - build Singularity and Docker image managers on top of that
- [ ] Singularity --request-singularity-image flag to use singularity to run bundle.
- [ ] Rest server change: include singularity image over 
- [ ] Frontend singularity changes (analogous to docker, use docker or singularity option)

**Comment by epicfaace** [comment]
- Do a POC: convert one of Tony's runs to an analogous Singularity version to ensure that it works
- Update design doc to include how the singularity workflow will work, as well as the steps for implementation
- How does Singularity work with GPUs (like nvidia-docker)?

**Comment by adiprerepa** [comment]
nvidia+singularity: https://sylabs.io/guides/3.5/user-guide/gpu.html

**Comment by adiprerepa** [comment]
I investigated the docker images that @jzwang43 found to not work by default, but this may be an issue with *singularity* and not us. A very fixable one at that.

Using the examples of mongodb and neo4j:

To reproduce (using the example of `mongo:latest`):


```sh
# build singularity image - this will create the file mongo.sif
singularity build mongo.sif docker://mongo:latest
# run singularity image - this produces an error
singularity run mongo.sif
```

Using this default implementation of singularity will produce an `$1 unbound variable` error - meaning the entrypoint is expecting an argument but is not being supplied one.

To see what this argument is, you can build the image with docker and see what docker passes to `docker-entrypoint.sh`:

```sh
# get mongo image and run container
docker pull mongo:latest && docker run -d mongo

# look for the actual command being run for the entrypoint - should say "docker-entrypoint.sh mongod"
docker ps --format '{{.Command}}' --no-trunc | grep mongod
```
The second command should output "docker-entrypoint.sh mongod". Note the `mongod` argument being passed to `docker-entrypoint.sh`.

If we look at the Dockerfile for [mongodb](https://github.com/dockerfile/mongodb/blob/master/Dockerfile#L25), on line 25, you will see `CMD ["mongod"]`. Whatever is in `CMD` is being passed to the `docker-entrypoint.sh`. In this case, it is "mongod".

If we go back to singularity and run the same commands as above but pass in `mongod` into singularity run:

```sh
singularity run mongo.sif mongod
```

The container runs fine. I confirmed this suspicion by doing the same with `neo4j:latest` ([dockerfile](https://github.com/neo4j/docker-neo4j/blob/master/docker-image-src/4.3/Dockerfile#L51), and adding `neo4j` to `singularity run` fixed the errors.

Singularity should handle passing in the contents of the docker `CMD` when converting docker->singularity images/entrypoints, not codalab.

**Comment by nelson-liu** [comment]
It seems like I got `mongod` to work without explicitly passing in the CMD:

```
# MongoDB looks for a mount at /data/ with data/db and data/configdb, so make that
mkdir -p data/db data/configdb
singularity run --bind data:/data docker://mongo:latest
```

**Comment by adiprerepa** [comment]
huh, this is weird
![image](https://user-images.githubusercontent.com/43913660/119069685-6fb2f280-b99b-11eb-9d52-9a4e9fa5361f.png)
I'm still getting the error.

**Comment by adiprerepa** [comment]
note to self - don't rely on apt to get you the latest version of singularity. the issue no longer exists

**Comment by adiprerepa** [comment]
POC is done: ran one of tony's runs locally overnight and it worked perfectly

**Comment by adiprerepa** [comment]
POC is done: ran one of tony's runs locally overnight and it worked perfectly

**Comment by adiprerepa** [comment]
![image](https://user-images.githubusercontent.com/43913660/119218184-edf2c000-ba93-11eb-9e65-a2e696d349dd.png)
using tensorflow as the gpu checker, it works!

**Comment by adiprerepa** [comment]
```bash
SINGULARITYENV_CUDA_VISIBLE_DEVICES=0,1 singularity run --nv tensorflow_latest-gpu.sif python3 foo.py 
```
foo.py:
```python3
from tensorflow.python.client import device_lib
print(device_lib.list_local_devices())
```

The environment variable `SINGULARITYENV_CUDA_VISIBLE_DEVICES` needs to be set to a list? I have to look into this

**Comment by adiprerepa** [comment]
the behavior in https://github.com/codalab/codalab-worksheets/blob/master/codalab/worker_manager/slurm_batch_worker_manager.py#L360 and our logic in requesting gpusets is the same, setting `SINGULARITYENV_CUDA_VISIBLE_DEVICES` to the list of indexes or gpu ids will change the visibility to the container.

**Comment by adiprerepa** [comment]
Look into how cpu/gpu bundles are directed to workers and mimimc

**Comment by adiprerepa** [comment]
https://github.com/codalab/codalab-worksheets/pull/3569 is only the first step. Jobs will not be run in the singularity container yet, in fact, nothing is really run in the singularity container yet.

Once that PR is merged, we need to create some other abstraction like `ImageRunner` or something that encapsulates the logic from `docker_utils.py` in functions like `start_bundle_container` and all other analogous functions.

The `ImageRunner` class (options of types `DockerImageRunner` and `SingularityImageRunner`) would run their respective images in `worker_run_state.py`.

We will also have to remove all the docker-specific logic from `worker_run_state.py` in the transition functions and all of the other functions and transfer it to the `ImageRunner` class.

**Comment by epicfaace** [comment]
@adiprerepa can you split this issue into multiple ones like https://github.com/codalab/codalab-worksheets/issues/3298?

**Comment by epicfaace** [comment]
Closing in favor of https://github.com/codalab/codalab-worksheets/issues/3661.
